components/x-interaction/src/Hydrate.jsx

```diff
@@ -1,27 +1,80 @@
+export class HydrationWrapper extends Component {
+	render() {
+		const {Component, props, id} = this.props;
+		return <Component {...props} id={id} actionsRef={a => this.actions = a} />;
+	}
+
+	componentDidMount() {
+		if(this.props.wrapper) {
+			this.props.wrapper.addEventListener('x-interaction.trigger-action', this);
+		}
+	}
+
+	componentWillUnmount() {
+		if(this.props.wrapper) {
+			this.props.wrapper.removeEventListener('x-interaction.trigger-action', this);
+		}
+	}
+
+	handleEvent(event) {
+		const {action, args = []} = event.detail;
+
+		if(this.actions && this.actions[action]) {
+			this.actions[action](...args);
+		}
+	}
+}
+
+export function hydrate() {
 	if (typeof window === 'undefined') {
 		throw new Error('x-interaction hydrate should only be called in the browser');
 	}
 
 	if (!('_xDashInteractionHydrationData' in window)) {
 		throw new Error(
-			"x-interaction hydrate was called without hydration data available. this could happen if you called hydrate before the hydration data was defined, or if you're not ouptutting the hydration data in your server-rendered markup."
+			"x-interaction hydrate was called without hydration data available. this could happen if you called hydrate above where the hydration data was defined on the page, or if you're not ouptutting the hydration data in your server-rendered markup."
 		);
 	}
 
+	const serialiserOrdering = `make sure you're always outputting the serialiser's data in the same request that the serialiser was created. see https://financial-times.github.io/x-dash/components/x-interaction/#hydrating for more details.`;
+
 	window._xDashInteractionHydrationData.forEach(({ id, component, props }) => {
-		const element = document.querySelector(`[data-x-dash-id="${id}"]`);
+		const wrapper = document.querySelector(`[data-x-dash-id="${id}"]`);
+
+		if(!wrapper) {
+			throw new Error(
+				`component markup for ${id} was not found on the page. It was expected to be an instance of ${component}. it's likely that this hydration data is from another request. ${serialiserOrdering}`
+			);
+		}
+
 		const Component = getComponent(component);
 
-		while (element.firstChild) {
-			element.removeChild(element.firstChild);
+		while (wrapper.firstChild) {
+			wrapper.removeChild(wrapper.firstChild);
 		}
 
-		render(<Component {...props} id={id} />, element);
+		render(<HydrationWrapper {...{
+			Component,
+			props,
+			id,
+			wrapper,
+		}} />, wrapper);
 	});
-}
 
-export default hydrate;
+	document.querySelectorAll('[data-x-dash-id]').forEach(element => {
+		const {xDashId} = element.dataset;
+
+		const hasData = !window._xDashInteractionHydrationData.find(
+			({ id }) => id === xDashId
+		);
+
+		if(!hasData) {
+			throw new Error(
+				`found component markup for ${xDashId} without any hydration data. it's likely that its hydration data has been output in another request, or that the component was rendered after the serialisation data was output. ${serialiserOrdering}`
```

Comment:
Should "hydration data" be "serialised data" or "dehydrated data" ?

Reply:
here's my reasoning: it's the component that's dehydrated, not the data, and the data is used to hydrate, so it's the hydration data? any old data can be serialised but this is for a specific purpose

Reply:
Got it. Maybe near the start of the doc it could lead into it, e.g.

> The serialised data used for hydrating a component (_hydration data_)